File: SORT-CHAR-BY-FREQUENCY/scbf.py
from collections import Counter

class Solution:
    def frequencySort(self, s: str) -> str:
        freq = Counter(s)
        sorted_items = sorted(freq.items(), key=lambda x: -x[1])
        print(sorted_items)


        result = []
        for char, count in sorted_items:
            result.append(char * count)

        return ''.join(result)

s = Solution()
s.frequencySort("tree")
s.frequencySort("cccaaa")
s.frequencySort("Aabb")
s.frequencySort("loveleetcode")

File: SORT-CHAR-BY-FREQUENCY/scbf.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

func frequencySort(s string) string {
	freq := make(map[rune]int)
	for _, ch := range s {
		freq[ch]++
	}

	type pair struct {
		char  rune
		count int
	}
	var pairs []pair
	for ch, cnt := range freq {
		pairs = append(pairs, pair{ch, cnt})
	}

	sort.Slice(pairs, func(i, j int) bool {
		return pairs[i].count > pairs[j].count
	})

	var sb strings.Builder
	for _, p := range pairs {
		sb.WriteString(strings.Repeat(string(p.char), p.count))
	}

	return sb.String()
}

func main() {
	s := "tree"
	result := frequencySort(s)
	fmt.Println(result) // Output: eert or ttee
}
